add integration server tests for resources version checks

endorama opened the pull request:
## Motivation/summary

<!--
Describe your change in the title and description, and provide a motivation for the
change and rationale for the approach taken.
-->

Add additional integrationservertest for a recent bug affecting some upgrade paths in APM Server.

## Checklist

<!--
Delete irrelevant items. The changelog should only be updated for user-facing changes.
Once the PR is ready for review there should be no unticked boxes.
-->

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

<!--
Explain how this PR can be tested by the reviewer: commands, dependencies, steps, etc.
If it is self-explanatory, delete this section.
-->

## Related issues

<!--
Reference the related issue(s), and make use of magic keywords where it makes sense
https://help.github.com/articles/closing-issues-using-keywords/.
-->

endorama committed:
add TypedClient()

endorama committed:
make sure terraform outputs are available

endorama committed:
allow skipping cleanup entirely

This is useful when you want to manually inspect the results even on success

endorama committed:
extract get deployment info

endorama committed:
add tests for bug

endorama committed:
with the new version there are more rollovers

endorama committed:
print when lazy rollover are expected to happen

This aid troubleshooting as you don't need to look into the upgrade-config.yaml
file but the indications are printed with the test

endorama committed:
support patch versions in HasLazyRollover, improve docs

Add support for specifying patch versions in upgrade-config.yaml.
If a patch version is present will take precedence over it's own
minor version, if both are specified.

Patch version use is supported in both keys and values. This allows
to specify, for example, specific patch versions not triggering a
lazy rollover when upgrading to another patch version.

Overhauled the comments in upgrade-config.yaml to reflect this new
possibilities and clarify how to write the configuration with examples.

endorama committed:
rework 8.18.3 to 8.19.0 test

Other than usual bug checks we need one more to
check for event.success_count

ericywl commented on integrationservertest/setup.go:
```diff
@@ -175,15 +175,28 @@ func createCluster(
-			t.Log("test failed and cleanup-on-failure is false, skipping cleanup")
-		}
-	})
+	if *doCleanup {
```
I don't think this works. The `doCleanup` is defined in `main_test.go` and so is only accessible in test files.

endorama replied:
It does if you run the tests 😅 It's a quick hack for my use case.

Do you think there is value in bringing this to `main`? If yes I'll do a proper implementation.

endorama commented on integrationservertest/apm-version-bug_test.go:
```diff
@@ -0,0 +1,180 @@
+			eventIngestedHasMappings.indexName = regexp.MustCompile(".ds-traces-apm-default-[0-9.]+-000003")
+		}
+
+		if len(versions) == 2 { // if 2 versions we expect the first to be bugged and the second to be not
+			return []testStep{
+				steps[0], // create
+				steps[1], // ingest
+				zeroEventIngestedDocs,
+				noEventIngestedInMappings,
+				steps[2], // upgrade
+				steps[3], // ingest
+				someEventIngestedDocs,
+				eventIngestedHasMappings,
+			}
+
+		} else if len(versions) == 3 { // if 3 versions we expect the first to be bugged on create, the second on upgrade, the third to be not
+			return []testStep{
+				steps[0], // create
+				steps[1], // ingest
+				steps[2], // upgrade
+				steps[3], // ingest
+				zeroEventIngestedDocs,
+				noEventIngestedInMappings,
+				steps[4], // upgrade
+				steps[5], // upgrade
+				someEventIngestedDocs,
+				eventIngestedHasMappings,
+			}
+		}
```
You are suggesting to not use the `buildTestStep` function but create them manually so it's easier to follow the various cycles?

ericywl replied:
My bad I wasn't clear enough. Creating them manually works, but we can also change how steps are defined in the runner. In the example above, `testCycle` is essentially just: `type testCycle []testStep`, and demarcates where one version ends and the other starts.

In your case, since you are inserting the steps in between each "cycle", it would be much less flimsy since you can inject the steps at the end of each cycle.

Just an idea, not sure if it would work well.

endorama committed:
fix 8.18.2 -> 8.18.3 test, add other notes

endorama committed:
refactor to clarify test steps

endorama committed:
fix 8.17.7 to.8.17.8

Finding event.ingested in 8.17.7 already after upgrade from 8.16.2
is expected due to the changes to apm-data between the versions.

endorama committed:
refactor version selection and lock versions

Refactor version selection to ease the move to BCs.
Don't use vsCache.GetLatestSnapshot as that will change
version as we release new ones. These tests are version specific
at the moment, so use fixed SNAPSHOT versions

endorama committed:
use BC/releases

Replace SNAPSHOT versions with BC or released versions.

Use SNAPSHOTS for 8.19.0 test as there is no BC version for it yet
and we cannot mix releases/BCs and SNAPSHOT versions.

endorama commented:
Closing as this dedicated to a specific debug session.
